helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+> [name=Matt Butcher]
+> Doing a little testing today, it does appear that if we render a chart and store the rendered output, that gives us sufficient information to do a full rollback. If this is the case, then we can retain rollback pretty easily, though the ReleaseVersion will have secrets data stored in it. That said, it's also _much_ easier to lock down with RBAC.
+
+#### Versions are ULIDs, not incremented integers
+Helm 2 used incremented integers for version numbering. This is changing.
+
+Version increments are ULIDs, not SemVer or integer counters. This is to prevent any race conditions when two revisions are pushed in near proximity. The version incrementor need not be negotiated with ULID; each new ULID is time-bound, but unique.
+
+
+## Hook Annotations
+
+In Helm 2, hook annotations could be placed on various manifests, and those annotations would result in special treatment.
+
+In Helm 2, resources with hook annotations were unmanaged. In Helm 3, the owner reference is set on those, and they will be removed when the chart is deleted. There will be a way to override the management policy by setting an annotation for `helm.sh/hook-policy: unmanaged`
+> [name=Taylor Thomas]
+> Would hooks be needed anymore with the new eventing system? I think all we would need is something that can translate Helm v2 hooks into the new events behind the scenes
+
+## Lua Plugins
+
+The existing Helm plugin mechanism has enabled numerous extensions to Helm (e.g., [here](https://docs.helm.sh/related/#helm-plugins) and [here](https://github.com/search?q=topic%3Ahelm-plugin&type=Repositories)). While this has been useful there have been two shortcomings:
+
+1. Many plugines are written as shell scripts assuming a POSIX environment and familiar linux-style userland. Helm, and Kubernetes as a whole, supports windows where these plugins don't work natively.
+2. Other plugins are written in a compiled languages (e.g., Go) or a general scripting language (e.g., Python) that have system dependencies in order to work. These can have version complications as well (e.g., a system Python of 3 and a plugin Python need of 2.7).
+
+Along with Helm supporting Lua for scripting chart extensions, it will support plugins in Lua. These plugins will only have a dependency on Helm to run and can be ported cross platform.
+> [name=Matt Butcher]
+> The main danger here is that we have to make it so that charts cannot have implicit reliance on Lua plugins. I think that is fairly straightforward, though. We just have to make sure that it's not possible to execute plugin code from within a chart.
+
+
+### Plugin Sandboxing
+
+Like charts, each plugin will be executed in an isolated Lua sandbox. Unlike chart scripts, plugins will be granted access to all core libraries, and will not require security prompts to accept.
```

Comment:
This worries me as an admin greatly....
Today you tend to run helm from your admin box. but this allows charts to reach back to your admin box and do just about anything. it can grab things, upload them to containers it manages, then ship them somewhere else, potentially obscuring the action in the process.

This means that kubernetes/charts becomes significantly less safe IMO.

Before helm was just uploading a templated set of k8s objects.

This change makes a helm package a turing complete application running on the admins box with access to a k8s cluster too.

Reply:
Not sure I follow. Plugins always had this security model. I'm just making it explicit that Lua plugins would have approximately the same model as the `exec`-style plugins.

The in-chart Lua scripts are not plugins, and are sandboxed differently.

Reply:
OH. I missed the distinction between plugins and chart scripts. Thanks.